_posts/2019-02-08-resize-images-on-the-fly.md

```diff
@@ -0,0 +1,187 @@
+
+## Create a new Node.js function
+
+So, let’s create our function to resize images, we’ll call it image-resize. Instead of the classic Node template, I’ll be using the node10-express template because it offers full control over HTTP headers and higher throughput. That means that we can set the Content-Type when resizing different types of images for our users.
+
+1. Create a new Function
+
+Create a new folder for our work
+
+```bash
+$ mkdir -p ~/dev/resizer \
+&& cd ~/dev/resizer
+```
+
+Now, pull the node10-express template from the OpenFaaS Template Store.
+
+```bash
+$ faas-cli template store pull node10-express
+```
+
+Create new function from the template
+
+```bash
+$ faas-cli new image-resize --lang=node10-express --prefix=USERNAME
+```
+
+> Note: Replace USERNAME, with your name on the Docker Hub, or your private registry address.
+
+This will create a new folder named image-resize having handler.js and package.json in it.
+
+```bash
+image-resize/handler.js
+image-resize/package.json
+image-resize.yml
+```
+
+The newly generated function is a hello-world style example which we can edit in a moment.
+
+Let’s deploy it and try it out.
+
+```bash
+$ faas-cli up -f image-resize.yml
+```
+
+We’ll check the deployed function. Then we will continue to write our image resizing functionality.
+
+(Image here)
+
+The default sample function serving traffic in the OpenFaaS Portal.
+
+2. Write the resizing logic using sharp
+
+We are going to use the sharp package to resize the images. This is where we would have normally used ImageMagick or GraphicsMagick. This is what the sharp website says about the library:
+
+> The typical use case for this high speed Node.js module is to convert large images in common formats to smaller, web-friendly JPEG, PNG and WebP images of varying dimensions.
+> Resizing an image is typically 4x-5x faster than using the quickest ImageMagick and GraphicsMagick settings.
+
+So at IconScout, having a 4x-5x speed increase was a great move forward for us where we process 10s of thousands of images per day and that’s growing every week.
+
+Now let’s install the sharp package using `npm`.
+
+```bash
+$ cd ~/dev/resizer/image-resizer
+npm install --save sharp
+```
+We’ll also need the request package to fetch the specified image from its URL.
+
+```bash
+$ cd ~/dev/resizer/image-resizer
+$ npm install --save request
+```
+
+Resizing images with sharp is really easy and simple to use, here is a link to the documentation: http://sharp.dimens.io/en/stable/api-resize/
+
+For example, to resize input.png to 200px x 200px, you only need the following code:
+
+```js
+sharp('input.png')
+  .png()
+  .resize(200, 200, {
+    fit: 'contain'
+  })
+  .toFile('output.png')
+```
+
+Let’s wrap this resizing code to our function.
+
+You can find the full code in this GitHub repo: https://github.com/tarunmangukiya/openfaas-functions/blob/master/image-resize/handler.js
```

Comment:
You should probably say something like, "copy this file your `image-resize/handler.js`"

Reply:
Actually, re-reading, you have this a couple lines down, on line  149. It might make sense to move that here so that the  source and the call to action are immediately next to each other.

Reply:
When I copied this content to my test function, the function fails to deploy 

```
Forking - node [index.js]
2019/02/06 14:44:08 Started logging stdout from function.
2019/02/06 14:44:08 Started logging stderr from function.
2019/02/06 14:44:08 OperationalMode: http
2019/02/06 14:44:08 Writing lock-file to: /tmp/.lock
2019/02/06 14:44:08 stderr: /home/app/function/node_modules/sharp/lib/libvips.js:61
      throw new Error(`'${vendorPlatformId}' binaries cannot be used on the '${currentPlatformId}' platform. Please remove the 'node_modules/sharp/vendor' directory and run 'npm install'.`);
      ^

2019/02/06 14:44:08 stderr: Error: 'darwin-x64' binaries cannot be used on the 'linuxmusl-x64' platform. Please remove the 'node_modules/sharp/vendor' directory and run 'npm install'.
    at Object.hasVendoredLibvips (/home/app/function/node_modules/sharp/lib/libvips.js:61:13)
    at
2019/02/06 14:44:08 stderr:  Object.<anonymous> (/home/app/function/node_modules/sharp/lib/constructor.js:9:22)
    at Module._compile (internal/modules/cjs/loader.js:688:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:699:10)
    at Module.load (internal/mod
2019/02/06 14:44:08 stderr: ules/cjs/loader.js:598:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:537:12)
    at Function.Module._load (internal/modules/cjs/loader.js:529:3)
    at Module.require (internal/modules/cjs/loader.js:636:17)
    at require (internal/modules/cjs/h
2019/02/06 14:44:08 stderr: elpers.js:20:18)
    at Object.<anonymous> (/home/app/function/node_modules/sharp/lib/index.js:3:15)

2019/02/06 14:44:08 Error reading stdout: EOF
```

This is obviously an issue with the fact that I am using a mac, but I am trying to figure out the correct sequence of commands to fix or avoid this. 

---

I fixed this by removing the `package-lock.json` that is generated when I run `npm install --save sharp`, per this line https://github.com/openfaas/openfaas.github.io/pull/66/files#diff-e80fade808999279ed88462ae23802d6R106

Reply:
Removing package-lock.json seems like an odd one.. why do we need to do that? Mismatch between our Mac and the Linux VM?

Reply:
I added this note